shivyc/il_cmds/math.py

```diff
@@ -118,6 +118,70 @@ class Mult(_AddMult):
     Inst = asm_cmds.Imul
 
 
+class _BitShiftCmd(ILCommand):
+    """Base class for bitwise shift commands."""
+
+    # The ASM instruction to generate for this command. Override this value
+    # in subclasses.
+    Inst = None
+
+    def __init__(self, output, arg1, arg2): # noqa D102
+        self.output = output
+        self.arg1 = arg1
+        self.arg2 = arg2
+
+    def inputs(self): # noqa D102
+        return [self.arg1, self.arg2]
+
+    def outputs(self): # noqa D102
+        return [self.output]
+
+    def abs_spot_pref(self): # noqa D102
+        return {self.arg2: [spots.RCX]}
+
+    def rel_spot_pref(self): # noqa D102
+        return {self.output: [self.arg1]}
+
+    def make_asm(self, spotmap, home_spots, get_reg, asm_code): # noqa D102
+        arg1_spot = spotmap[self.arg1]
+        arg1_size = self.arg1.ctype.size
+        arg2_spot = spotmap[self.arg2]
+        arg2_size = self.arg2.ctype.size
+
+        # According Intel® 64 and IA-32 software developer's manual
+        # Vol. 2B 4-582 second (count) operand must be represented as
+        # imm8 or CL register.
+        if not self._is_imm8(arg2_spot) and arg2_spot != spots.RCX:
```

Comment:
Aaah, I just realized the potential for a pesky bug here. It's possible arg1_spot will be RCX, meaning that a command moving arg2 into RCX will overwrite the value of arg1 before we use it. Could you add a check for this situation? In pseudocode, something like:
```
if arg2 isn't an immediate and is not in RCX:    # you have this line already
   if arg1 is in RCX:
      move arg1 to a temporary register (preferably output_spot, definitely not RCX or arg2 spot)
   move arg2 to RCX    # you have this code already
```
Writing IL commands is tricky because there's lots of ways subtle bugs like this can creep in, and they won't come up without lots of testing or thinking really hard about all the possible situations.